x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/mapper/Mapper.java

```diff
@@ -170,7 +121,13 @@ private PhysicalPlan mapUnary(UnaryPlan unary) {
         }
 
         if (unary instanceof Limit limit) {
-            mappedChild = addExchangeForFragment(limit, mappedChild);
+            // Avoid extra limit for remote Enrich, since it has already been applied before Enrich if it's duplicate.
```

Comment:
The extra limit is fine. It'll be optimized away on the data node. (limits are pushed past enrichs.) I'd leave this the data node's responsibility, otherwise the code might get brittle.

(We have no way to prove that this duplicated limit isn't somehow still crucial. Other optimizations may have picked up the duplicate limit and done something with it. In fact, it's not incorrect for other optimizations to have removed the duplicate limit before we reach here.)

Reply:
In the case of Enrich, it's not crucial because we will have a copy of it pre-enrich. It's not true for other operations (like joins) because they can expand cardinality. So the mapper does the right thing in the general case, but in specifically cause of Enrich this limit is useless. I mainly inserted this to not have to redo a bunch of tests by inserting extra limit, but if you think it's fine we could leave it alone and let the data node optimizer handle it (I'll check if it does).